storage/src/main/java/org/commonjava/storage/pathmapped/core/PathMappedFileManager.java

```diff
@@ -187,11 +195,21 @@ public long getFileLastModified( String fileSystem, String path )
 
     public boolean exists( String fileSystem, String path )
     {
-        if ( path == null )
+        // NOS-2289 After checking the database, I found that there are some mismatch between database entries
+        //          and physical store files. There are entries in database, but the corresponding physical files
+        //          missed in physical location. So I think we need to check real existence of from physical level
+        //          but not just from db level to make sure if it really exists.
+        if ( StringUtils.isBlank( path ) )
         {
+            logger.warn( "Checking existence of an empty path for file system {}, which is considered as not existed", fileSystem );
             return false;
         }
-        return pathDB.exists( fileSystem, path );
+        if ( isDirectory( fileSystem, path ) )
+        {
+            return pathDB.exists( fileSystem, path );
+        }
+        final String storageFile = pathDB.getStorageFile( fileSystem, path );
+        return physicalStore.exists( storageFile );
```

Comment:
I am afraid this will hide the real issue and slow down the method because exists is called very often. We might just let it fail and figure out the real problem. I suspect there was some accident. Some days ago I found the quary.io indy-stage-static was linked to devel (or vise verse), which might have caused some confusion. If it worked on the old Indy code and removed some files (or metadata files?) and switched to path-mapped later, there might be some files changed under the old /maven folder, and cause mismatches in db and disk. I suspect this should not have happened to new folders. Our code simply inserts the db when/after file stream is closed. Another thing is the file GC, but I really doubt it could cause this.

Reply:
That's why I add this type of real check of existence, because we should always to keep the exists result same with the following other file operations to make sure the whole file ops process is right. If it is treated as existed, but the following file opening reported it as not existed, that will always make this exists api very confused. We really don't know what will happen in the future which could cause this type of mismatch in the future, like system crash or anything else, so I think to make sure this consistence is important.
For example with this bug, it is caused by the mismatch of a npm package.json, which could be fetched from remote if it does not exist on local. But because of this mismatch, indy thinks it exists on the system so by pass the fetch from the remote, but when do the real reading ops by open stream, it reported a null because the physical file is not there. That's why this NPE happened. From upper system this type of error is really confusing because it can not give correct judgement based on exits method because it does not give the right answer for the following file operations.